lib/util.py

```diff
@@ -678,79 +679,135 @@ class TxHashMismatch(ServerError):
 
 import socket
 import ssl
+import errno
+
+class JSONSocketPipe(PrintError):
+    """ Non-blocking wrapper for a socket passing one-per-line json messages:
+
+       <json><newline><json><newline><json><newline>...
+
+    Correctly handles SSL sockets and gives useful info for select loops.
+    """
 
-class SocketPipe(PrintError):
-    class MessageSizeExceeded(RuntimeError):
-        ''' Raised by get() if max_message_bytes is set and the message size
-        limit was exceeded. '''
+    class Closed(RuntimeError):
+        ''' Raised if socket is closed '''
 
     def __init__(self, socket, *, max_message_bytes=0):
         ''' A max_message_bytes of <= 0 means unlimited, otherwise a positive
         value indicates this many bytes to limit the message size by. This is
         used by get(), which will raise MessageSizeExceeded if the message size
         received is larger than max_message_bytes. '''
         self.socket = socket
-        self.message = b''
-        self.set_timeout(0.1)
+        socket.settimeout(0)
         self.recv_time = time.time()
         self.max_message_bytes = max_message_bytes
-
-    def set_timeout(self, t):
-        self.socket.settimeout(t)
+        self.recv_buf = bytearray()
+        self.send_buf = bytearray()
 
     def idle_time(self):
         return time.time() - self.recv_time
 
-    def clean_up(self):
-        ''' Clears the receive buffer to make sure no garbage data remains '''
-        self.message = b''
+    def get_selectloop_info(self):
+        ''' Returns tuple:
+
+        read_pending - new data is available that may be unknown to select(),
+            so perform a get() regardless of select().
+        write_pending - some send data is still buffered, so make sure to call
+            send_flush if writing becomes available.
+        '''
+        try:
+            # pending() only defined on SSL sockets.
+            has_pending = self.socket.pending() > 0
+        except AttributeError:
```

Comment:
Old C++ habit -- using exceptions for normal control flow in a 10Hz loop seems just sinful. I actually don't know how "expensive" exceptions are in Python as compared to normal control flow (the logic and stack unwinding is done itself in C so it's probably just as slow as normal execution right?).  

What are your thoughts on also adding a big `if` around this for whether the socket is SSL or not (maybe latch a boolean flag to true in c'tor if SSL)?  Or does it not matter anyway since in Python raising is equally slow to normal execution?  Hmm.. Just curious what you think.

Reply:
Heh I did feel naughty writing that. :-) In python it's generally seen as pythonic to do this kind of thing - EAFP vs LBYL. I like the idea of first snagging a boolean in the constructor though.

Reply:
Up to you.  On second thought in Python it's probably just as slow as normal execution to `raise` -- since the exception handling/stack unwinding is done in C so it's fast-ish compared to normal Python execution anyway. Hmm.  Up to you.  This has the advantage of just being far far clearer/easier to read/less state.